src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+	for (int i = 0; i < samples_to_consume; ++i)
+		dest[i] = convert_float_to_int16(cd->output[i]);
+	dest = audio_stream_wrap(sink, dest + samples_to_consume);
+	output_buffers->size += audio_stream_frame_bytes(sink) * samples_to_consume / n_ch;
+
+	cd->output_ptr -= samples_to_consume;
+	for (int i = 0; i < cd->output_ptr; ++i)
+		cd->output[i] = cd->output[i + samples_to_consume];
+}
+
+static void ctc_s24_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s24(source, src));
+	int samples_to_consume = 0;
+
+	for (int i = 0; i < samples_to_process; ++i)
+		cd->input[cd->input_ptr++] = convert_int32_to_float(src[i]);
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+
+	while (cd->input_ptr >= cd->num_frames * n_ch) {
+		GoogleCtcAudioProcessingProcess(cd->state, cd->input, cd->output + cd->output_ptr,
+						cd->num_frames, n_ch);
+		cd->output_ptr += cd->num_frames * n_ch;
+
+		cd->input_ptr -= cd->num_frames * n_ch;
+		for (int i = 0; i < cd->input_ptr; ++i)
+			cd->input[i] = cd->input[i + cd->num_frames * n_ch];
+	}
+
+	samples_to_consume = MIN(cd->output_ptr, audio_stream_samples_without_wrap_s24(sink, dest));
+	for (int i = 0; i < samples_to_consume; ++i)
+		dest[i] = convert_float_to_int32(cd->output[i]);
+	dest = audio_stream_wrap(sink, dest + samples_to_consume);
+	output_buffers->size += audio_stream_frame_bytes(sink) * samples_to_consume / n_ch;
+
+	cd->output_ptr -= samples_to_consume;
+	for (int i = 0; i < cd->output_ptr; ++i)
+		cd->output[i] = cd->output[i + samples_to_consume];
+}
+
+static void ctc_s32_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s32(source, src));
+	int samples_to_consume = 0;
+
+	for (int i = 0; i < samples_to_process; ++i)
+		cd->input[cd->input_ptr++] = convert_int32_to_float(src[i]);
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
```

Comment:
It suddenly strikes me that this all almost 100% identical to the same code in AEC, though this has 24 bit sample support.  Post-merge we should do a unification pass.

Reply:
+1